Zend/zend_compile.c

```diff
@@ -2867,8 +2891,17 @@ static void zend_compile_list_assign(
 
 		zend_verify_list_assign_target(var_ast, old_style);
 
-		zend_emit_op(&fetch_result, ZEND_FETCH_LIST, expr_node, &dim_node);
-		zend_emit_assign_znode(var_ast, &fetch_result);
+		opline = zend_emit_op(&fetch_result, ZEND_FETCH_LIST, expr_node, &dim_node);
+
+		if (elem_ast->attr || (var_ast->kind == ZEND_AST_ARRAY && zend_set_list_assign_reference(var_ast))) {
+			opline->extended_value = ZEND_LIST_MAKE_WRITABLE;
+		}
+
+		if (elem_ast->attr) {
+			zend_emit_assign_ref_znode(var_ast, &fetch_result);
+		} else {
+			zend_emit_assign_znode(var_ast, &fetch_result);
+		}
```

Comment:
However, constants, and tmpvars, appear to be caught during execution raising a Fatal, wherein assignments can only happen to writable values.  So it's preventing the odd uses.  I assume the check you're asking for here is a compile time check, rather than an execution based one.